test/sql/cte/recursive_cte_key_variant.test

```diff
@@ -0,0 +1,195 @@
+# name: test/sql/cte/recursive_cte_key_variant.test
+# description: Recursive CTEs with an error thrown in the pipelines
+# group: [cte]
+
+statement ok
+PRAGMA enable_verification;
+
+query III
+WITH RECURSIVE tbl(a,b,c) USING KEY (a) AS (SELECT * FROM (VALUES (1, 2, 3), (2,2,5)) UNION SELECT a+1, b+1, c FROM tbl WHERE b < 4) SELECT * FROM  tbl;
+----
+1	2	3
+2	3	3
+3	4	3
+4	4	5
+
+query III
+WITH RECURSIVE tbl(a,b,c) USING KEY (b) AS (SELECT * FROM (VALUES (1, 2, 3), (2,3,2)) UNION SELECT a, b+1, c FROM tbl WHERE b < 4) SELECT * FROM  tbl;
+----
+1	2	3
+1	3	3
+1	4	3
+
+query III
+WITH RECURSIVE tbl(a,b,c) USING KEY (a) AS (SELECT * FROM (VALUES (1, 2, 3), (2,3,2)) UNION SELECT a, b+1, c FROM tbl WHERE b < 4) SELECT * FROM  tbl;
+----
+1	4	3
+2	4	2
+
+query III
+WITH RECURSIVE tbl(a,b,c) USING KEY (b,c) AS (SELECT * FROM (VALUES (1, 2, 3), (2,3,3)) UNION SELECT a, b+1, c FROM tbl WHERE b < 4) SELECT * FROM  tbl;
+----
+1	2	3
+1	3	3
+1	4	3
+
+query III
+WITH RECURSIVE tbl(a,b,c) USING KEY (a,b,c) AS (SELECT * FROM (VALUES (1, 2, 3), (2,3,3)) UNION SELECT a, b+1, c FROM tbl WHERE b < 4) SELECT * FROM  tbl;
+----
+1	2	3
+2	3	3
+1	3	3
+2	4	3
+1	4	3
+
+query III
+WITH RECURSIVE tbl(a,b,c) USING KEY (a) AS (SELECT * FROM (VALUES (1, 2, 3), (1,3,3)) UNION SELECT a, b+1, c FROM tbl WHERE b < 4) SELECT * FROM  tbl;
+----
+1	4	3
+
+query III
+WITH RECURSIVE tbl(a,b,c) USING KEY (a,b,c) AS (SELECT * FROM (VALUES (1, 2, 3), (1,3,3)) UNION SELECT a, b+1, c FROM tbl WHERE b < 4) SELECT * FROM  tbl;
+----
+1	2	3
+1	3	3
+1	4	3
+
+# Reference the recurring table more than once
+query III
+WITH RECURSIVE tbl(a,b,c) USING KEY (a) AS (SELECT 1, NULL, NULL UNION SELECT tbl.a+1, rec1.a, rec2.b FROM tbl, recurring_tbl AS rec1, recurring_tbl AS rec2 WHERE tbl.a < 5) SELECT * FROM tbl;
+----
+1	NULL	NULL
+2	1	NULL
+3	2	1
+4	3	2
+5	4	3
+
+
+#######################
+# Connected components
+#######################
+statement ok
+CREATE OR REPLACE TABLE nodes (node int);
+
+statement ok
+INSERT INTO nodes VALUES (0), (1), (2), (3), (4), (5), (6);
+
+statement ok
+CREATE OR REPLACE TABLE edges (here int, there int);
+
+statement ok
+INSERT INTO edges VALUES  (0,4), (4,0),
+						  (0,3), (3,0),
+						  (1,4), (4,1),
+						  (3,4), (4,3),
+						  (2,5), (5,2);
+
+
+
+query II
```

Comment:
Can we perhaps include a larger benchmark comparing the standard recursive CTEs vs the new ones as part of the benchmark runner (see e.g. `benchmark/micro/cast/cast_date_string.benchmark`)?

Reply:
I will write more test, when the syntax is determined. :+1: